### DISCUSSION TITLE
Exploring Apprise instead of maintaining (n)th notification channels

### DISCUSSION THREAD

**Discussion by alexjustesen** [alexjustesen/speedtest-tracker#1961] [Polls]
In short, I'm entertaining the idea of dropping all notification channels except database, email and webhooks in favor of integrating with [Apprise](https://github.com/caronc/apprise) for notifications.

This discussion and poll is meant to evaluate this decision as it would be considered a deviation and breaking change from the current state. Feel free to use the poll below as a quick yay/nay to voice your opinion and if you have further context please leave a comment as well.

### 👍 Pros

- Less notification channels I have to support which frees up capacity to work on features.
- 100+ supported notification channels: https://github.com/caronc/apprise?tab=readme-ov-file#supported-notifications

### 👎 Cons

- Will require separate setup of Apprise by users to integrate with services like Slack, Telegram and Discord.
- tbd...

Thanks in advance and I'm looking forward to hearing everyone's thoughts.

**Comment by pigers** [comment]
Maybe implement Shoutrrr ?

**Reply by alexjustesen** [reply]
I had considered it but since Shoutrrr was "inspired by Apprise" it made more sense to consider the more popular project.

**Reply by pigers** [reply]
yeah - if there's more rich project, then picking it up makes more sense !

**Reply by pigers** [reply]
had  a look at apprise, that would require to provide good installation and configuration routine for users
it's documentation is confusing to read.

**Comment by armond-avanes** [comment]
I'm a heavy Apprise/Apprise-API user so I might be a bit biased.

I wouldn't call it separate setup for the existing notifications. All they have to do is to migrate their existing setup from something like
`https://discord.com/api/webhooks/WEBHOOK_ID/WEBHOOK_TOKEN` to `discord://WEBHOOK_ID/WEBHOOK_TOKEN` (and almost the same for Slack and other channels). And by the way, the migration process can be fully (or at least partially) automated as part of SpeedTest-Tracker version migration and the users won't even notice the change. Usually the URL formats are pretty standard and solid. They can be parsed and migrated to the new format.

The good thing here is that, by just migrating to this new setup format, you're going to get access to 100 more channels under the same roof as you've already mentioned! This is a great feature to have both for you as the project maintainer (less code to implement and less docs to maintain) and for the users to get instant access to several more channels which are not supported now.

UPDATE: Here is a consolidated list of supported notifications and their formats as part of Apprise docs: https://github.com/caronc/apprise/#supported-notifications

**Reply by svenvg93** [reply]
I did some research for the apprise integration while working on it, so far I could see there is no nice way to integrate their CLI into the app. At least not without adding python to the image which is not really preferred. The only I could see is to use their docker container as a "sidecar" and sent the notification to that one. 

Do you know by any chance any ways to implement the CLI without adding dependencies  to manage?

**Reply by armond-avanes** [reply]
Looking into your Docker file (https://github.com/alexjustesen/speedtest-tracker/blob/main/docker/8.3/Dockerfile), you're already installing `python2` package there (line # 25). But even if it's not meant to be there, I'm not sure what issues it can cause by adding python runtime to your image. It's just a package there and being used when only needed (Apprise CLI in this case).

If the docker image size is the main concern, you can consider migrating to an Alpine based image instead of running on full ubuntu version.

**Reply by armond-avanes** [reply]
I'm not a PHP developer at all, but doing a quick search reveals that it's pretty straightforward to call a python script from PHP. Here is one of those quick guides: https://www.tutorialspoint.com/run-a-python-program-from-php

Shouldn't be much of a problem to implement this integration.

**Reply by svenvg93** [reply]
This Docker file is only used for the development environment. The container is build by [lsio](https://github.com/linuxserver/docker-speedtest-tracker) which doesn't have python  in their image. 

> I'm not a PHP developer at all, but doing a quick search reveals that it's pretty straightforward to call a python script from PHP. Here is one of those quick guides: https://www.tutorialspoint.com/run-a-python-program-from-php
> 
> Shouldn't be much of a problem to implement this integration.

It's indeed not a problem to call it from the application. The main "problem" is the extra dependencies. 

@alexjustesen What are your thoughts on it?

**Reply by armond-avanes** [reply]
> It's indeed not a problem to call it from the application. The main "problem" is the extra dependencies.

I understand your concern. But those extra dependencies are just kind of an isolated runtime and being utilized only when needed (and when Apprise CLI is called). They're not going to affect the main/existing PHP application, unless I'm missing something here.

**Reply by pigers** [reply]
why not create a stack of apprise and speedtest in yaml and call it a day ? other apps are having mariadb as part of stack required to run software and it's totally fine.

**Reply by armond-avanes** [reply]
> why not create a stack of apprise and speedtest in yaml and call it a day ? other apps are having mariadb as part of stack required to run software and it's totally fine.

For me, it would be perfectly fine. I'm already running a standalone Apprise-API container to consolidate all my notifications config there.

But I doubt that people not familiar with Apprise ecosystem will find this useful. It's not even a database which makes sense to a major part of the user base. Imagine having to setup a second container (and thus forcing to use Docker Compose instead of plain Docker) just for the sake of sending notifications. I still prefer an internal integration of Apprise CLI within the main docker image. Much simpler and straightforward from users' perspective. Just my 2 cents.

**Comment by armond-avanes** [comment]
@alexjustesen @svenvg93 Any final decision on Apprise integration?

**Reply by alexjustesen** [reply]
Sven has done some awesome work but at the moment I'm tapped out time wise with other projects (need to pay the bills). Speedtest Tracker is in maintenance mode for the time being and hopefully circle back around to new features in the spring.

**Reply by armond-avanes** [reply]
Quite understandable. I truly appreciate the time and efforts you and Sven have been putting into this side project. Thanks for the update.

**Comment by crispybegs** [comment]
is there a way of getting slack notifications? i have all my services reporting in there. an old version of speedtest i was using used to do slack, but that eventually disappeared

**Reply by svenvg93** [reply]
Slack Webhooks are available in the notification settings.

**Reply by crispybegs** [reply]
> Slack Webhooks are available in the notification settings.

ah ok, I can see there's a webhook option but nothing slack specific. I tried entering a slack webhook but, when testing the notification, nothing was received. Am i getting something wrong?

Edit: scratch that. i think i was running an old version. just updated and can see slack now!

**Comment by Chucky2401** [comment]
I use Healthcecks notification service.
Will it be supported with the Webhook notification service?

**Reply by svenvg93** [reply]
Webhooks are just generic [payload](https://docs.speedtest-tracker.dev/settings/notifications/webhook). Im not sure if these will work for Healthchecks. Would say give it a try and let us know how it went.

**Reply by alexjustesen** [reply]
@Chucky2401 I used Healthchecks.io through the webhook channel so yes.

**Comment by lnlyssg** [comment]
I can see the Apprise integration is live but it needs tag support! All my other services use tags and I configure all the actual services in Apprise itself, that way if I decided to switch notification services I only have to update Apprise.

**Reply by alexjustesen** [reply]
> I can see the Apprise integration is live but it needs tag support!

Can you hit us with some docs on apprise tags? We're happy to add them if this would be a useful part of the feature.

Edit: Also if you can open an issue to add tag support that would be idea, discussions I don't monitor as closely.

**Reply by lnlyssg** [reply]
Done #2503

**Comment by devanteweary** [comment]
I just noticed this and I'm confused...

So now I have to set up yet another completely separate app if I want to continue to get notifications in Discord? 
Why remove options instead of adding Apprise as another option? Current Discord notifications are perfect.

I tried converting the Discord webhook URL to the discord: URL but seems like it also needs an Apprise server URL if I'm not looking at this wrong.

I just want to verify before I get into learning how to set up yet another app if I want to continue notifications. 
It's just working so well now I don't get why removing it.
